controllers/dnspolicy_dnsrecords.go

```diff
@@ -20,6 +20,43 @@ import (
 	"github.com/kuadrant/kuadrant-operator/pkg/multicluster"
 )
 
+type DNSMetricsEventHandler struct{}
+
+func (m *DNSMetricsEventHandler) OnCreate(_ context.Context, obj client.Object, err error) {
+	if err != nil {
+		return
+	}
+	dnsRecord, ok := obj.(*kuadrantdnsv1alpha1.DNSRecord)
+	if !ok {
+		return
+	}
+	dnsRecordEvent.WithLabelValues(dnsRecord.Name, dnsRecord.Namespace, "create").Inc()
```

Comment:
This combo of labels feels odd for create & delete.
Will the counter only be incremented 1 time ever as the record with that name & namespace combo only be created or deleted once?

Maybe a better option is to have an '_info' like metric be created/deleted (and have a value of 1) when the record is created/deleted.
See `kube_pod_info` in https://github.com/kubernetes/kube-state-metrics/blob/main/docs/metrics/workload/pod-metrics.md for an example

Reply:
I was under the impression that doing this would allow a query like: "show me how many dns record created events there are every 5 seconds for the last day", so it would highlight for example, 2 hours ago there was a huge jump in create and delete events.

Reply:
Because of the name & namespace label being included, you'll end up with a new time series for each dnsrecord.
Each series will never increment beyond 1.
For example, if you create 3 new records (record1, record2, record3), you'll get 3 new time series like this:

```
dns_record_event{name="record1",namespace="whatever"} 1
dns_record_event{name="record2",namespace="whatever"} 1
dns_record_event{name="record3",namespace="whatever"} 1
```

Reply:
wouldn't it be:
```
dns_record_event{name="record1",namespace="whatever", event="create"} 1
dns_record_event{name="record2",namespace="whatever", event="create"} 1
dns_record_event{name="record3",namespace="whatever", event="create"} 1
```

Reply:
Yes.
Is this what you want?

From my understanding on how time series data is stored in prometheus, this would mean there's a new time series with the new label combination each time a new record is created and scraped for the first time. 
If you just want a count on creates, a single time series may be sufficient where the values are what's saved each scrape.

More info in https://www.robustperception.io/cardinality-is-key/

Reply:
I suppose what I want is 2 things:
- be able to see a spike in creates or deletes
  - Determine if it's lots of unique records, or the same record deleting and recreating over and over
- see a spike in updates
  - Identify which records are updating a lot.

If you have recommendations on how I could best formulate this metric to achieve, I'm happy to follow any advice!

Reply:
I wonder if you can satisfy those use cases sufficiently with existing controller runtime metrics https://book.kubebuilder.io/reference/metrics-reference
Here's a screenshot from a dev cluster showing the rate of reconciles, grouped by controller (so, by resource).
![image](https://github.com/user-attachments/assets/73e35c96-15ce-4a43-a7a3-3bb2c80d9599)
and the query
```
sum(rate(controller_runtime_reconcile_total{job="dns-operator-controller-manager-metrics-service"}[5m])) by(controller)
```
If the rate of reconciles is "too high", an alert could fire.
You'd have to jump to logs to get higher resolution data on what's causing the spike.
Alertnatively, you go the route you are thinking and have that higher resolution data (dnsrecord names & namespaces) as labels on the time series data and have users take the potential storage hit.

There may be other useful labels in those controller runtime metrics as well, like `result` to show if the reconciles are mostly successes or errors. It might be worth exploring them if you haven't already.
Here's a screenshot of what that kind of grouping might look like
<img width="1661" alt="image" src="https://github.com/user-attachments/assets/0ae31615-cc90-4691-ba8b-6af3f73898ea">

Reply:
ok, then I think it makes sense to remove these event metrics for now, which also significantly simplifies this PR, and as you say, use the controller_runtime metrics to identify the spike in activity is coming from Kuadrant-operator, and use logs to work out what it's doing.

That sounds like a feasible solution to me, which also allows me to remove all the event handler code from the base_reconciler.